### PR TITLE
CLI Improvements

### LINKED ISSUE #62: Add real cmdline flag parsing
Get `argparse` involved, use nice cmdline error messages, spit out version numbers automagically, etc.

### PR DESCRIPTION
**Pull request by newsch**
With the shiny new cli, you can: 
- continue to connect to a server 
- continue to load from a file or stdin 
- specify a height/width for the canvas, (which will also resize it after loading a file) 
- get an automagic `git describe` version 
- file an issue on github 
- get an introduction to the interface 

I made the current version with `argp`, but that's GNU C and @labseven had some trouble compiling it on macOS. The options are:
- continue using argp, and either find a way of using it on macOS or choose not to support it
- switch to getopt, which is less nice but standard C
- make our own or use a 3rd party library
  - https://github.com/cofyc/argparse
  - https://github.com/argtable/argtable3 or http://argtable.sourceforge.net/

TODO:
- [x] pick one of the above methods
- [x] validate arguments
- [x] add note about default canvas size
- [x] remove test print comments
- [x] rebase and squash before merge

Closes #62

### CODE REVIEW

**Comment by newsch** [comment]
Right now `argtable` looks like the best option - cross-platform, offers a better interface than `getopt`, and would be easy to integrate (just a single `.c` and `.h` file).

**Comment by newsch** [comment]
As [mentioned here](https://stackoverflow.com/questions/9642732/parsing-command-line-arguments-in-c#comment69442798_24479532), `argp` can also be included in the project through [gnulib](https://www.gnu.org/software/gnulib/). This is my new favorite option.

**Comment by newsch** [comment]
Side note: [`libiberty` is a pretty funny name](https://gcc.gnu.org/onlinedocs/libiberty/).

**Comment by newsch** [comment]
> As [mentioned here](https://stackoverflow.com/questions/9642732/parsing-command-line-arguments-in-c#comment69442798_24479532), `argp` can also be included in the project through [gnulib](https://www.gnu.org/software/gnulib/). This is my new favorite option.

Nevermind, it needs autoconf and would add 142 new files and 35k LOC.

File: src/canvas_test.c [hunk 1]
```diff
@@ -179,11 +179,11 @@ MU_TEST(test_canvas_cpy_p1p2) {
 
   c4 = canvas_cpy_p1p2(c3, 1, 1, c3->num_rows - 1, c3->num_cols - 1);
 
-  printf("\nc3\n");
-  canvas_print(c3);
+  // printf("\nc3\n");
```
Comment (MatthewBeaudouinLafon) [inline src/canvas_test.c]: remove comments?

File: src/frontend.c [hunk 2]
```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
...
+  bool connect_remote;  // connect to a remote server
+  char *remote_host;
+  char *remote_port;
+} arguments_t;
+
+/* Make sure the arguments struct is valid and in a non-conflicting state.
+ *
+ * Return values are meant for an argp parser function.
+ *
+ * TODO: fill this out
```
Comment (MatthewBeaudouinLafon) [inline src/frontend.c]: Either do this or make an issue?

File: src/frontend.c [hunk 3]
```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
...
+#endif
+
+// cli pieces
+const char *program_name = "collascii";
+const char *program_version = VERSION;
+const char *program_bug_address = "<https://github.com/olin/collascii/issues/>";
+const char *program_description = "a collaborative ascii editor";
+const char *program_doc =
+    "COLLASCII\n\n"
+    "\"The Future Editor of Yesterday, Tomorrow!\"\n\n"
```
Comment (MatthewBeaudouinLafon) [inline src/frontend.c:105]: Might be good to have a word here on starting a server (for someone who hasn't gone through extensive collascii training). Something like:

> To start a collascii server, make and run server.out with `make server.out && ./server.out`

File: src/frontend.c [hunk 4]
```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
...
+    exit(0);
+  }
+
+  if (usage->count > 0) {
+    printf("Usage: %s", program_name);
+    arg_print_syntax(stdout, argtable, "\n");
+    exit(0);
+  }
+
+  if (version->count > 0) {
```
Comment (MatthewBeaudouinLafon) [inline src/frontend.c:199]: running `./collascii --version --usage` only prints the usage due to the `exit(0)` in the usage block. I don't think it's a problem, just checking that it's intended behavior.

File: src/frontend.c [hunk 5]
```diff
@@ -82,7 +90,210 @@ char *logfile_path = "out.txt";
...
+  struct arg_str *server, *port;
+  struct arg_end *end;
+
+  void *argtable[] = {
+      help = arg_litn(NULL, "help", 0, 1, "display this help and exit"),
+      usage =
+          arg_litn(NULL, "usage", 0, 1, "display brief usage info and exit"),
+      version =
+          arg_litn(NULL, "version", 0, 1, "display version info and exit"),
+      width = arg_intn("w", "width", "<n>", 0, 1, "initial width of canvas"),
```
Comment (MatthewBeaudouinLafon) [inline src/frontend.c]: Is it worth saying that 0 width goes to default value?

**Review by MatthewBeaudouinLafon — COMMENTED** [review]
A few minor comments, but it looks good otherwise. Happy to approve once you've taken a look.

File: src/frontend.c (hunk 3 above)
Comment (newsch) [reply src/frontend.c:105]: Would the README be a better place for that?

File: src/frontend.c (hunk 4 above)
Comment (newsch) [reply src/frontend.c:199]: The behavior I've seen from other programs is `--usage`, `--version`, and `--help` are all mutually exclusive. I tried both `--version` and `--usage` on `man` and collascii built with `argp`, and both of them returned the results for only the first one invoked - I don't know of a way to do that here though, I don't think `argtable` can return their positions.

One option would be to return an error if both are invoked.

File: src/frontend.c (hunk 5 above)
Comment (newsch) [reply src/frontend.c]: I'm inclined to leave out the 0 bit, include a message about the default canvas size, and reword the width and height args to say they resize/set a custom value.

File: src/frontend.c (hunk 3 above)
Comment (MatthewBeaudouinLafon) [reply src/frontend.c:105]: That too. I'm thinking that someone who just downloaded collascii after skimming the Readme might do `--help` and like to see that info there.

File: src/frontend.c (hunk 4 above)
Comment (MatthewBeaudouinLafon) [reply src/frontend.c:199]: Alright, sounds like not worth the effort then. Let's stick to current behavior.

File: src/frontend.c (hunk 3 above)
Comment (newsch) [reply src/frontend.c:105]: I added some changes the README and the CLI.

File: src/frontend.c (hunk 2 above)
Comment (newsch) [reply src/frontend.c]: Done!

**Comment by newsch** [comment]
> Happy to approve once you've taken a look.

Rebased and ready!

**Review by MatthewBeaudouinLafon — APPROVED** [review]

**Pull request merged**